cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/templates/etl/api/batch/BatchContext.java

```diff
@@ -0,0 +1,105 @@
+import co.cask.cdap.api.RuntimeContext;
+import co.cask.cdap.api.data.DatasetContext;
+import co.cask.cdap.api.data.batch.BatchReadable;
+import co.cask.cdap.api.data.batch.InputFormatProvider;
+import co.cask.cdap.api.data.batch.OutputFormatProvider;
+import co.cask.cdap.api.data.batch.Split;
+import co.cask.cdap.api.dataset.Dataset;
+
+import java.util.List;
+
+/**
+ * Context passed to Batch Sources and Sink.
+ */
+public interface BatchContext extends RuntimeContext, DatasetContext {
+
+  /**
+   * Returns the logical start time of the Batch Job.  Logical start time is the time when this MapReduce
+   * job is supposed to start if this job is started by the scheduler. Otherwise it would be the current time when the
+   * job runs.
+   *
+   * @return Time in milliseconds since epoch time (00:00:00 January 1, 1970 UTC).
+   */
+  long getLogicalStartTime();
+
+  /**
+   */
+  <T> T getHadoopJob();
+
+  /**
+   * Overrides the input configuration of this Batch job to use
+   * the specified dataset by its name.
+   *
+   * @param datasetName the name of the input dataset.
+   */
+  void setInput(String datasetName);
+
+  /**
+   * Overrides the input configuration of this Batch job to use
+   * the specified dataset by its name and data selection splits.
+   *
+   * @param datasetName the name of the input dataset
+   * @param splits the data selection splits
+   */
+  void setInput(String datasetName, List<Split> splits);
+
+  /**
+   * Overrides the input configuration of this MapReduce job to write to the specified dataset instance.
+   *
+   * <p>
+   * Currently, the dataset passed in must either be an {@link InputFormatProvider} or a {@link BatchReadable}.
+   * You may want to use this method instead of {@link #setInput(String, List)} if your input dataset uses runtime
+   * arguments set in your own program logic. Input splits are determined either from the dataset instance if it is
+   * a {@link BatchReadable} or from the corresponding MapReduce InputFormat if the dataset instance is a
+   * {@link InputFormatProvider}.
+   * </p>
+   *
+   * @param datasetName the name of the input dataset
+   * @param dataset the input dataset
+   */
+  void setInput(String datasetName, Dataset dataset);
+
+  /**
+   * Overrides the output configuration of this Batch job to write to the specified dataset by its name.
+   *
+   * @param datasetName the name of the output dataset
+   */
+  void setOutput(String datasetName);
+
+  /**
+   * Overrides the output configuration of this Batch job to write to the specified dataset instance.
+   * Currently, the dataset passed in must either be an {@link OutputFormatProvider}.
+   * You may want to use this method instead of {@link #setOutput(String)} if your output dataset uses runtime
+   * arguments set in your own program logic.
+   *
+   * @param datasetName the name of the output dataset
+   * @param dataset the output dataset
+   */
+  void setOutput(String datasetName, Dataset dataset);
```

Comment:
So the BatchContext allows someone to set both input and output?? Then what's the point of having BatchSource and BatchSink??